src/utils/command_manager.h

```diff
@@ -64,20 +66,30 @@ class command_manager : public ::dsn::utils::singleton<command_manager>
         command_handler handler;
     };
 
+    void deregister_command(uintptr_t handle);
+
     typedef ref_ptr<command_instance> command_instance_ptr;
     utils::rw_lock_nr _lock;
     std::map<std::string, command_instance_ptr> _handlers;
 };
 
-} // namespace dsn
+class command_deregister
+{
+public:
+    command_deregister(uintptr_t id) : cmd_id_(id) {}
+    ~command_deregister()
+    {
+        if (cmd_id_ != 0) {
+            dsn::command_manager::instance().deregister_command(cmd_id_);
+            cmd_id_ = 0;
+        }
+    }
 
-#define UNREGISTER_VALID_HANDLER(ptr)                                                              \
-    do {                                                                                           \
-        if (ptr != nullptr) {                                                                      \
-            dsn::command_manager::instance().deregister_command(ptr);                              \
-            ptr = nullptr;                                                                         \
-        }                                                                                          \
-    } while (0)
+private:
+    uintptr_t cmd_id_ = 0;
```

Comment:
Can `cmd_id_` be type of `command_instance *`, or just `std::shared_ptr<command_instance>` ?